src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs

```diff
@@ -3432,6 +3476,486 @@ public void Inheritance_17()
+        public void Deconstruct_MethodCollision_03()
+        {
+            var source = @"
+using System;
+
+record B
+{
+    public int X() => 3;
+}
+
+record C(int X, int Y) : B
+{
+    public new int X { get; }
+
+    static void M(C c)
+    {
+        switch (c)
+        {
+            case C(int x, int y):
+                Console.Write(x);
+                Console.Write(y);
+                break;
+        }
+    }
+
+    static void Main()
+    {
+        M(new C(1, 2));
+    }
+}
+";
+            var verifier = CompileAndVerify(source, expectedOutput: "02");
+            verifier.VerifyDiagnostics();
+
+            Assert.Equal(
+                "void C.Deconstruct(out System.Int32 X, out System.Int32 Y)",
+                verifier.Compilation.GetMember("C.Deconstruct").ToTestDisplayString(includeNonNullable: false));
+        }
+
+        [Fact]
+        public void Deconstruct_MethodCollision_04()
+        {
+            var source = @"
+record C(int X, int Y)
+{
+    public int X(int arg) => 3;
+
+    static void M(C c)
+    {
+        switch (c)
+        {
+            case C(int x, int y):
+                break;
+        }
+    }
+
+    static void Main()
+    {
+        M(new C(1, 2));
+    }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (2,14): error CS0102: The type 'C' already contains a definition for 'X'
+                // record C(int X, int Y)
+                Diagnostic(ErrorCode.ERR_DuplicateNameInClass, "X").WithArguments("C", "X").WithLocation(2, 14));
+
+            Assert.Equal(
+                "void C.Deconstruct(out System.Int32 X, out System.Int32 Y)",
+                comp.GetMember("C.Deconstruct").ToTestDisplayString(includeNonNullable: false));
+        }
+
+        [Fact]
+        public void Deconstruct_Empty()
+        {
+            var source = @"
+record C
+{
+    static void M(C c)
+    {
+        switch (c)
+        {
+            case C():
+                break;
+        }
+    }
+
+    static void Main()
+    {
+        M(new C());
+    }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (8,19): error CS1061: 'C' does not contain a definition for 'Deconstruct' and no accessible extension method 'Deconstruct' accepting a first argument of type 'C' could be found (are you missing a using directive or an assembly reference?)
+                //             case C():
+                Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, "()").WithArguments("C", "Deconstruct").WithLocation(8, 19),
+                // (8,19): error CS8129: No suitable 'Deconstruct' instance or extension method was found for type 'C', with 0 out parameters and a void return type.
+                //             case C():
+                Diagnostic(ErrorCode.ERR_MissingDeconstruct, "()").WithArguments("C", "0").WithLocation(8, 19));
+
+            Assert.Null(comp.GetMember("C.Deconstruct"));
```

Comment:
Where do we decide to skip generating a `Deconstruct()`?

Reply:
We only enter the code path to generate Deconstruct when the record declaration has a parameter list.